rust/lance-index/src/scalar/inverted.rs

```diff
@@ -0,0 +1,513 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use arrow::array::{AsArray, ListBuilder, UInt64Builder};
+use arrow::datatypes;
+use arrow_array::{ArrayRef, RecordBatch, StringArray, UInt32Array, UInt64Array};
+use arrow_schema::{DataType, Field};
+use async_trait::async_trait;
+use datafusion::execution::SendableRecordBatchStream;
+use deepsize::DeepSizeOf;
+use futures::{StreamExt, TryStreamExt};
+use itertools::Itertools;
+use lance_core::{Error, Result, ROW_ID};
+use roaring::RoaringBitmap;
+use snafu::{location, Location};
+
+use crate::Index;
+
+use super::{IndexReader, IndexStore, ScalarIndex, ScalarQuery};
+
+const TOKENS_FILE: &str = "tokens.lance";
+const INVERT_LIST_FILE: &str = "invert.lance";
+const DOCS_FILE: &str = "docs.lance";
+
+const TOKEN_COL: &str = "_token";
+const TOKEN_ID_COL: &str = "_token_id";
+const FREQUENCY_COL: &str = "_frequency";
+const NUM_TOKEN_COL: &str = "_num_tokens";
+
+#[derive(Debug, Clone, Default, DeepSizeOf)]
+pub struct InvertedIndex {
+    tokens: TokenSet,
+    invert_list: InvertedList,
+    docs: DocSet,
+}
+
+impl InvertedIndex {
+    // map tokens to token ids
+    // ignore tokens that are not in the index cause they won't contribute to the search
+    fn map(&self, texts: &[String]) -> Vec<u32> {
+        texts
+            .iter()
+            .filter_map(|text| self.tokens.get(text))
+            .collect()
+    }
+}
+
+#[async_trait]
+impl Index for InvertedIndex {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn as_index(self: Arc<Self>) -> Arc<dyn Index> {
+        self
+    }
+
+    fn as_vector_index(self: Arc<Self>) -> Result<Arc<dyn crate::vector::VectorIndex>> {
+        Err(Error::invalid_input(
+            "inverted index cannot be cast to vector index",
+            location!(),
+        ))
+    }
+
+    fn statistics(&self) -> Result<serde_json::Value> {
+        Ok(serde_json::json!({
+            "num_tokens": self.tokens.tokens.len(),
+            "num_docs": self.docs.row_ids.len(),
+        }))
+    }
+
+    fn index_type(&self) -> crate::IndexType {
+        crate::IndexType::Scalar
+    }
+
+    async fn calculate_included_frags(&self) -> Result<RoaringBitmap> {
+        unimplemented!()
+    }
+}
+
+#[async_trait]
+impl ScalarIndex for InvertedIndex {
+    // return the row ids of the documents that contain the query
+    async fn search(&self, query: &ScalarQuery) -> Result<UInt64Array> {
+        let row_ids = match query {
+            ScalarQuery::FullTextSearch(texts) => {
+                let tokens = self.map(texts);
+                let row_ids = tokens
+                    .iter()
+                    .filter_map(|token| self.invert_list.retrieve(*token))
+                    .flat_map(|(row_ids, _)| row_ids.iter().cloned())
```

Comment:
Can this lead to duplicate row ids?  E.g. if row 0 contains "a b" and `texts` is `["a", "b"]` then will row 0 appear twice in the results?